File: src/lowlevel/server/icefsoperators/icefsFlush.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 16:24:55
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsFlush.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsFlush(fh uint64) (status int32) {
	newFh, err := syscall.Dup(int(fh))
	status = icefserror.IcefsStdErrno(err)
	if err == nil {
		syscall.Close(newFh)
	}
	return
}

func (s *IcefsGRpcServer) DoIcefsFlush(ctx context.Context, req *pb.IcefsFlushReq) (*pb.IcefsFlushRes, error) {
	var res pb.IcefsFlushRes

	res.Status = s.server.doIcefsFlush(req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsFlush(ctx context.Context, req *icefsthrift.IcefsFlushReq) (*icefsthrift.IcefsFlushRes, error) {
	var res icefsthrift.IcefsFlushRes

	res.Status = s.server.doIcefsFlush(uint64(req.Fh))

	return &res, nil
}

File: src/lowlevel/client/icefsoperators/icefsFlush.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:43
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsFlush.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsFlush(fuse_req_t fuseReq, fuse_ino_t inode,
                               struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsFlushReq req;
      icefsgrpc::IcefsFlushRes res;
      grpc::ClientContext ctx;

      req.set_fh(fi->fh);

      grpc::Status status = gRpcClient->DoIcefsFlush(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsFlushReq req;
      icefsthrift::IcefsFlushRes res;

      req.__set_fh(fi->fh);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsFlush(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}
